perception/navigator_vision/nodes/database_color.py

```diff
@@ -108,21 +119,90 @@ def publish(self, *args):
                 rospy.logerr(e)
 
 
+class Observation(object):
+    history_length = 100  # Default to 100
+
+    def __init__(self):
+        self.hues = deque([], maxlen=self.history_length)
+        self.values = deque([], maxlen=self.history_length) 
+        self.q_errs = deque([], maxlen=self.history_length)
+        self.dists = deque([], maxlen=self.history_length)
+    
+    def __len__(self):
+        return len(self.hues)
+   
+    def __iadd__(self, (hue, value, dist, q_diff)):
+        """
+        Add observations like:
+            >>> co += [0.65, 0.4, ..., 0.1]
+        """
+        self.hues.append(hue)
+        self.values.append(value)
+        self.dists.append(dist)
+        self.q_errs.append(q_diff)
+
+        return self
+  
+    def __repr__(self):
+        _str = 'hues: {}\nvalues: {}\ndists: {}\nq_errs: {}'
+        return _str.format(*np.round(map(np.array, [self.hues, self.values, self.dists, self.q_errs]), 3))
+
+    def extend(self, (hues, values, dists, q_diffs)):
+        """Add lists of data in at once"""
+        self.hues.extend(hues)
+        self.values.extend(values)
+        self.dists.extend(dists)
+        self.q_errs.extend(q_diffs)
+
+    def compute_confidence(self, (value_w, dist_w, q_diff_w), **kwargs):
+        """Don't try to compute weights with bad data (too few samples)"""
+
+        # Get some value information
+        v_u = kwargs.get('v_u', 230)
+        v_sig = kwargs.get('v_sig', 30)
+        dist_sig = kwargs.get('dist_sig', 70)
+        q_sig = kwargs.get('q_sig', 1.3)
+
+        # Compute data required before applying weights
+        value_errs = self._guass(self.values, v_u, v_sig)
+        dists = self._guass(self.dists, 5, dist_sig)
+        q_diffs = self._guass(self.q_errs, 0, q_sig)
+
+        # Normalize weights
+        w_norm = value_w + dist_w + q_diff_w
+        value_w /= w_norm
+        dist_w /= w_norm
+        q_diff_w /= w_norm
+
+        #print "1", hue_std_w, np.round(hue_std, 2), hue_std_w * hue_std
+        #print "2", value_w, np.round(value_errs, 2), value_w * value_errs
+        #print "3", dist_w, np.round(dists, 2), dist_w * dists
+        #print "4", q_diff_w, np.round(q_diffs, 2), q_diff_w * q_diffs
+
+        # Compute normalized confidence
+        c = value_w * value_errs + dist_w * dists + q_diff_w * q_diffs
+        return c
+    
+    def _guass(self, data, mean, sig):
+        return np.exp(-np.power(np.array(data).astype(np.float64) - mean, 2) / (2 * np.power(sig, 2)))
+
+
 class Colorama(object):
     def __init__(self):
         info_topic = camera_root + "/camera_info"
-        image_topic = camera_root + "/image_raw"  # Change this to rect on boat
+        image_topic = camera_root + "/image_rect_color_throttle"
```

Comment:
That message filters thing isn't implemented for python. If you think it will run okay with not throttled frames on boat please update it to use /image_raw